File: src/shared/observability/logger.ts
export interface Logger {
	info(message: string, meta?: Record<string, unknown>): void;
	warn(message: string, meta?: Record<string, unknown>): void;
	error(message: string, meta?: Record<string, unknown>): void;
}

function write(
	level: "INFO" | "WARN" | "ERROR",
	message: string,
	meta?: Record<string, unknown>,
): void {
	const payload = {
		level,
		message,
		timestamp: new Date().toISOString(),
		...meta,
	};

	const output = JSON.stringify(payload);
	if (level === "ERROR") {
		console.error(output);
		return;
	}

	console.log(output);
}

export function createLogger(): Logger {
	return {
		info(message, meta) {
			write("INFO", message, meta);
		},
		warn(message, meta) {
			write("WARN", message, meta);
		},
		error(message, meta) {
			write("ERROR", message, meta);
		},
	};
}

File: rust/crates/application/src/investigation/logger.rs
use std::collections::BTreeMap;

pub trait InvestigationLogger: Send + Sync {
    fn info(&self, message: &str, meta: BTreeMap<String, String>);
    fn warn(&self, message: &str, meta: BTreeMap<String, String>);
    fn error(&self, message: &str, meta: BTreeMap<String, String>);
}
